### PR TITLE
Better context info in `Type_Error` raised from return type checks

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Followup to #8502 that adds better error messages

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by radeusgd** [comment]
Example:
```
from Standard.Base import all

go x -> Integer =
    if x == 0 then 1 else
        if x == 10 then "TXT" else
            @Tail_Call go x-1

main =
    IO.println <| go 4
    IO.println <| go 12
```
yields:
```
1
Execution finished with an error: Type error: expected `go` result to be Integer, but got Text.
        at <enso> test.go(test.enso:3-6)
        at <enso> test.main(test.enso:10:5-23)
```

**Comment by radeusgd** [comment]
This is a simple approach to ensuring better error messages - when creating the type ascriptions that enforce the type checks, we add a 'comment' to it that will then drive the error message.

I was thinking of implementing a slightly cleaner approach where instead of adding the comment to the IR, we could deduce this in an IR pass from context (detecting the type check at the return of a function), but this seemed like a much simpler solution. 

Still I can try to do the 'better' one if we think this is not good. 

### Alternative solution

I think the toher solution may need then to actually change the overall appproach to type checks - remove the `Type.Ascriptions` added for return checks, instead adding some return type information to the IR of particular Functions or Methods. Then that type info can be used in a separate ReturnType pass that would generate the `ReadArgumentCheckNode` with proper comments.

---

I think regardless of this PR, I will try to experiment a bit with the current approach and see if this approach works well with the next steps of most basic type inference/propagation/checking. If I see that actually there the 'Alternative Solution' would be more desirable anyway, that may mean we should take it and scratch this PR - that's why I separated it from the base PR.

File: engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java [hunk 1]
```diff
@@ -944,13 +944,13 @@ public void returnTypeCheckOptInErrorZeroArgumentsExpression() throws Exception
...
-    assertEquals(8, plusChecked.execute(2).asInt());
+    var foo = module.invokeMember(MethodNames.Module.EVAL_EXPRESSION, "foo");
+    assertEquals(8, foo.execute(2).asInt());
     try {
-      var res = plusChecked.execute(".");
+      var res = foo.execute(".");
       fail("Expecting an exception, not: " + res);
     } catch (PolyglotException e) {
-      assertContains("expected `expression` to be Integer, but got Text", e.getMessage());
+      assertContains("expected `x` result to be Integer, but got Text", e.getMessage());
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: Shouldn't this be `expecting 'x' to return Integer, but got Text`?

**Review by JaroslavTulach — APPROVED** [review]
Computing a _"comment"_ automatically and attaching it to `IR` is fine. There is #8419 to start an effort of checking what's the fastest thing for `runtime-compiler` to do.

File: engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java (hunk 1 above)
Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: Yes I think that form may be more readable, I will update it.

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeSignatures.scala [hunk 2]
```diff
@@ -322,7 +324,8 @@ case object TypeSignatures extends IRPass {
     *
     * @param signature the expression for the type signature
     */
```
Comment (Akirathan) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeSignatures.scala:327]:
```suggestion
    * @param comment an optional comment from which the pontential error message will be derived
    */
```

File: engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java [hunk 3]
```diff
@@ -872,7 +872,7 @@ public void returnTypeCheckOptInError() throws Exception {
       var res = plusChecked.execute("a", "b");
       fail("Expecting an exception, not: " + res);
     } catch (PolyglotException e) {
-      assertContains("expected `expression` to be Integer, but got Text", e.getMessage());
+      assertContains("expected `plusChecked` result to be Integer, but got Text", e.getMessage());
```
Comment (Akirathan) [inline engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: What about the following error message:
```
expected the result of `plusChecked` method to be Integer, but got Text
```
Isn't that even better?

**Review by Akirathan — APPROVED** [review]
This definitely feels like a better error message for an unexpected return type from a method. I have even a better suggestion, I think.

File: engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java (hunk 3 above)
Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: sure it sounds good to me, I will amend when I will get around to moving this PR forward, thanks for the suggestion

Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: I just think that maybe `function` may be a more universal name here?

`method` fits well for `Type.foo` methods. But internal helper functions e.g.:
```
foo x =
    go i = ...
    go 0
```
are not really _method_s, are they?

So maybe 
```
expected the result of `plusChecked` function to be Integer, but got Text
```
?

Comment (Akirathan) [reply engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: I don't know. I thought that everything *function-like* in Enso is just called *method*.

**Review by GregoryTravis — APPROVED** [review]

**Review by JaroslavTulach — APPROVED** [review]

File: engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java (hunk 3 above)
Comment (radeusgd) [reply engine/runtime/src/test/java/org/enso/interpreter/test/SignatureTest.java]: I did
```
expected the result of `plusChecked` to be Integer, but got Text
```

avoiding the method/function controversy.

IMO it is readable.

**Pull request merged**